From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 16 May 2017 00:59:48 -0400
Subject: [PATCH 069/102] PaX shadow cr4 sanity check (essentially a revert)

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
[levente@leventepolyak.net: Adapt to cpu_tlbstate moved out-of-line]
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
[thibaut.sautereau@ssi.gouv.fr: Move BUG_ON from native_flush_tlb_global() to
				new __native_tlb_flush_global() helper]
Signed-off-by: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
---
 arch/x86/include/asm/tlbflush.h | 1 +
 arch/x86/kernel/cpu/common.c    | 1 +
 arch/x86/kernel/process.c       | 1 +
 3 files changed, 3 insertions(+)

diff --git a/arch/x86/include/asm/tlbflush.h b/arch/x86/include/asm/tlbflush.h
index 0545fe75c3fa1..d37c738cb23aa 100644
--- a/arch/x86/include/asm/tlbflush.h
+++ b/arch/x86/include/asm/tlbflush.h
@@ -508,9 +508,10 @@ extern void enter_lazy_tlb(struct mm_struct *mm, struct task_struct *tsk)
 	__compiletime_error("enter_lazy_tlb() should not be used in modules");
 #endif /* !MODULE */
 
 static inline void __native_tlb_flush_global(unsigned long cr4)
 {
+	BUG_ON(cr4 != __read_cr4());
 	native_write_cr4(cr4 ^ X86_CR4_PGE);
 	native_write_cr4(cr4);
 }
 #endif /* _ASM_X86_TLBFLUSH_H */
diff --git a/arch/x86/kernel/cpu/common.c b/arch/x86/kernel/cpu/common.c
index a3df21d264606..c0dbce3dd76d3 100644
--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -495,10 +495,11 @@ EXPORT_SYMBOL_GPL(native_write_cr4);
 #endif
 
 void cr4_update_irqsoff(unsigned long set, unsigned long clear)
 {
 	unsigned long newval, cr4 = this_cpu_read(cpu_tlbstate.cr4);
+	BUG_ON(cr4 != __read_cr4());
 
 	lockdep_assert_irqs_disabled();
 
 	newval = (cr4 & ~clear) | set;
 	if (newval != cr4) {
diff --git a/arch/x86/kernel/process.c b/arch/x86/kernel/process.c
index 85435044e33cd..d3231770401d6 100644
--- a/arch/x86/kernel/process.c
+++ b/arch/x86/kernel/process.c
@@ -705,10 +705,11 @@ void speculation_ctrl_update_current(void)
 }
 
 static inline void cr4_toggle_bits_irqsoff(unsigned long mask)
 {
 	unsigned long newval, cr4 = this_cpu_read(cpu_tlbstate.cr4);
+	BUG_ON(cr4 != __read_cr4());
 
 	newval = cr4 ^ mask;
 	if (newval != cr4) {
 		this_cpu_write(cpu_tlbstate.cr4, newval);
 		__write_cr4(newval);
-- 
2.55.0

